src/System.Drawing.Common/src/System/Drawing/Icon.Unix.cs

```diff
@@ -709,77 +709,163 @@ private void InitFromStreamWithSize(Stream stream, int width, int height)
             if (stream.Length == 0)
                 throw new System.ArgumentException("The argument 'stream' must be a picture that can be used as a Icon", nameof(stream));
 
-            //read the icon header
-            BinaryReader reader = new BinaryReader(stream);
+            bool sizeObtained = false;
+            bool valid = false;
+            // read the icon header
+            using (var reader = new BinaryReader(stream))
+            {
+                // iconDir = new IconDir ();
+                iconDir.idReserved = reader.ReadUInt16();
+                if (iconDir.idReserved != 0) //must be 0
+                    throw new System.ArgumentException("Invalid Argument", nameof(stream));
+
+                iconDir.idType = reader.ReadUInt16();
+                if (iconDir.idType != 1) //must be 1
+                    throw new System.ArgumentException("Invalid Argument", nameof(stream));
+
+                ushort dirEntryCount = reader.ReadUInt16();
+                imageData = new ImageData[dirEntryCount];
+                iconDir.idCount = dirEntryCount;
+                iconDir.idEntries = new IconDirEntry[dirEntryCount];
+
+                // now read in the IconDirEntry structures
+                for (int i = 0; i < dirEntryCount; i++)
+                {
+                    IconDirEntry ide = new IconDirEntry
+                    {
+                        width = reader.ReadByte(),
+                        height = reader.ReadByte(),
+                        colorCount = reader.ReadByte(),
+                        reserved = reader.ReadByte(),
+                        planes = reader.ReadUInt16(),
+                        bitCount = reader.ReadUInt16(),
+                        bytesInRes = reader.ReadUInt32(),
+                        imageOffset = reader.ReadUInt32()
+                    };
+                    iconDir.idEntries[i] = ide;
+
+                    long streamPosition = stream.Position;
+                    stream.Seek(iconDir.idEntries[i].imageOffset, SeekOrigin.Begin);
+                    byte[] buffer = new byte[iconDir.idEntries[i].bytesInRes];
+                    stream.Read(buffer, 0, buffer.Length);
+                    using (var bihReader = new BinaryReader(new MemoryStream(buffer)))
+                    {
+                        uint headerSize = bihReader.ReadUInt32();
 
-            //iconDir = new IconDir ();
-            iconDir.idReserved = reader.ReadUInt16();
-            if (iconDir.idReserved != 0) //must be 0
-                throw new System.ArgumentException("Invalid Argument", nameof(stream));
+                        // Vista 256x256 icons points directly to a PNG bitmap
+                        // 256x256 icons are decoded as 0x0 (width and height are encoded as BYTE)
+                        // and we ignore them just like MS does (at least up to fx 2.0)
```

Comment:
This comment can be updated -- we do not ignore them (from looking at Icon.Windows.cs, it is not clear that the loop has any means to skip PNG
https://github.com/dotnet/corefx/blob/cec9910fcaa0321cd33b6fd43924a2794c5f3bb3/src/System.Drawing.Common/src/System/Drawing/Icon.Windows.cs#L501-L570

It is this code that ignores them. ie., on Unix we do not support PNG in icons.

Reply:
Actually if I read right, we don't cmopletely ignore them. The `IconDump` is used when saving.

Reply:
we don’t ignore them, we just don’t parse them and store the raw data in order to be able to save.

Reply:
So "we ignore them just like MS does" is doubly wrong comment :)

Reply:
Agree. Will update.